File: generic_tools/cross_validation.js
import { indent, findAll, extractFirst, stringToUtf8Bytes, toRepresentation } from "https://deno.land/x/good@1.4.4.2/string.js"

const _ = (await import('https://cdn.skypack.dev/lodash@4.17.21'))

const makeIterable = (generator, length)=>{
    generator[Symbol.iterator] = ()=>generator()
    Object.defineProperties(generator, {
        length: {
            get() { return length }
        }
    })
    // generator.length = length
    return generator
}

/**
 * Perform cross-validation by partitioning the data into all possible combinations.
 * 
 * @example
 *     const numberOfFolds = crossValidation({
 *         inputs: [
 *             [1, 2],
 *             [2, 3],
 *             [3, 4],
 *             [4, 5],
 *             [5, 6],
 *         ],
 *         outputs: [0, 0, 1, 1, 1],
 *         numberOfFolds: 3,
 *     })
 *     console.log(foldIndices)
 *     // the "inputs" and "outputs" below 
 *     // are generators rather than arrays
 *     [
 *       {
 *         train: { inputs: [...], outputs: [...] },
 *         test:  { inputs: [...], outputs: [...] },
 *       },
 *       {
 *         train: { inputs: [...], outputs: [...] },
 *         test:  { inputs: [...], outputs: [...] },
 *       },
 *       {
 *         train: { inputs: [...], outputs: [...] },
 *         test:  { inputs: [...], outputs: [...] },
 *       }
 *     ]
 * @param {Array<Array>} inputs - The samples [0]= first sample input
 * @param {Array} outputs 
 * @param {number} numberOfFolds - The number of numberOfFolds for cross-validation.
 * @returns {Array<Object>} An array of fold objects containing train and test data indices.
 */
export function crossValidation({inputs, outputs, numberOfFolds, shouldRandomize=true}) {
    const numberOfSamples = inputs.length
    const foldSize = Math.floor(numberOfSamples / numberOfFolds)
    const folds = []
    if (typeof numberOfFolds != "number") {
        throw Error(`crossValidation({ numberOfFolds: }), numberOfFolds needs to be a number, instead I got: ${toRepresentation(numberOfFolds)}`)
    }

    const indicies = shouldRandomize ? _.shuffle([...Array(numberOfSamples)].map((e,i)=>i)) : [...Array(numberOfSamples)].map((e,i)=>i)

    for (let i = 0; i < numberOfFolds; i++) {
        const start = i * foldSize
        const end = i === numberOfFolds - 1 ? numberOfSamples : (i + 1) * foldSize
        const trainIndices = []
        const testIndices = []

        const indiciesCopy = [...indicies]
        
        for (let j = 0; j < numberOfSamples; j++) {
            if (j >= start && j < end) {
                testIndices.push(indiciesCopy.pop())
            } else {
                trainIndices.push(indiciesCopy.pop())
            }
        }
        
        // 
        // uses iterators to avoid memory usage and up-front computation
        // 
        folds.push({
            train: {
                inputs: makeIterable(function *() {
                    for (const each of trainIndices) {
                        yield inputs[each]
                    }
                },trainIndices.length),
                outputs: makeIterable(function *() {
                    for (const each of trainIndices) {
                        yield outputs[each]
                    }
                },trainIndices.length)
            },
            test: {
                inputs: makeIterable(function *() {
                    for (const each of testIndices) {
                        yield inputs[each]
                    }
                }, testIndices.length),
                outputs: makeIterable(function *() {
                    for (const each of testIndices) {
                        yield outputs[each]
                    }
                }, testIndices.length),
            },
        })
    }

    return folds
}

File: generic_tools/cross_validation.py
import math
from random import random, sample, choices, shuffle

def cross_validation(*data, number_of_folds, should_randomize=True):
    number_of_samples = len(inputs)
    fold_size = math.floor(number_of_samples / number_of_folds)
    folds = []
    
    indicies = list(range(number_of_samples))
    if should_randomize: shuffle(indicies)
    
    for index in range(number_of_folds):
        start = index * fold_size
        end = number_of_samples if index == number_of_folds - 1 else (index + 1) * fold_size
        train_indices = []
        test_indices = []
        
        copy_of_indicies = list(indicies)
        
        for j in range(number_of_samples):
            if j >= start and j < end:
                test_indices.append(copy_of_indicies.pop())
            else:
                train_indices.append(copy_of_indicies.pop())
        
        # 
        # uses iterators to avoid memory usage and up-front computation
        # 
        folds.append({
            "train": tuple(
                tuple(each_source[each_index] for each_index in train_indices)
                    for each_source in data
            ),
            "test": tuple(
                tuple(each_source[each_index] for each_index in test_indices)
                    for each_source in data
            ),
        })

    return folds